packages/edit-site/src/components/editor/global-styles-provider.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * WordPress dependencies
+ */
+import { createContext, useContext, useEffect } from '@wordpress/element';
+import { useEntityProp } from '@wordpress/core-data';
+
+/**
+ * Internal dependencies
+ */
+import { getGlobalStyles } from './global-styles-renderer';
+
+const GlobalStylesContext = createContext( {
+	/* eslint-disable no-unused-vars */
+	getProperty: ( context, family, name ) => {},
+	setProperty: ( context, family, name, value ) => {},
+	globalContext: {},
+	/* eslint-enable no-unused-vars */
+} );
+
+export const useGlobalStylesContext = () => useContext( GlobalStylesContext );
+
+export default ( { children, baseStyles, contexts } ) => {
+	const {
+		userStyles,
+		getProperty,
+		setProperty,
+	} = useGlobalStylesFromEntities();
+
+	useGlobalStylesEffectToUpdateStylesheet( contexts, baseStyles, userStyles );
+
+	return (
+		<GlobalStylesContext.Provider
+			value={ {
```

Comment:
On each component render we will provide a different object reference as value, probably causing lots of rerenders on the tree even if no user styles changes happened. Maybe we should some memorization etc to ensure we always pass the same reference unless there is a change.